integrationtest/mapper/orm-realbackend/src/test/java/org/hibernate/search/integrationtest/mapper/orm/realbackend/multitenant/RealBackendDatabaseMultitenancyIT.java

```diff
@@ -0,0 +1,160 @@
+ */
+package org.hibernate.search.integrationtest.mapper.orm.realbackend.multitenant;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.List;
+import java.util.function.Consumer;
+import javax.persistence.Entity;
+import javax.persistence.Id;
+
+import org.hibernate.Session;
+import org.hibernate.SessionFactory;
+import org.hibernate.search.integrationtest.mapper.orm.realbackend.testsupport.BackendConfigurations;
+import org.hibernate.search.mapper.orm.Search;
+import org.hibernate.search.mapper.pojo.mapping.definition.annotation.GenericField;
+import org.hibernate.search.mapper.pojo.mapping.definition.annotation.Indexed;
+import org.hibernate.search.util.common.SearchException;
+import org.hibernate.search.util.impl.integrationtest.common.FailureReportUtils;
+import org.hibernate.search.util.impl.integrationtest.mapper.orm.OrmSetupHelper;
+
+import org.junit.AssumptionViolatedException;
+import org.junit.Rule;
+import org.junit.Test;
+
+public class RealBackendDatabaseMultitenancyIT {
+
+	public static final String TENANT_ID_1 = "TENANT 1";
+	public static final String TENANT_ID_2 = "TENANT 2";
+	public static final String TENANT_TEXT_1 = "I'm in the TENANT 1";
+	public static final String TENANT_TEXT_2 = "I'm in the TENANT 2";
+
+	@Rule
+	public OrmSetupHelper setupHelper = OrmSetupHelper.withSingleBackend( BackendConfigurations.simple() );
+
+	private SessionFactory sessionFactory;
+
+	@Test
+	public void multiTenancyStrategy_discriminator() {
+		sessionFactory = setupHelper.start()
+				.withProperty( "hibernate.search.backend.multi_tenancy.strategy", "discriminator" )
+				.withProperty( "hibernate.search.automatic_indexing.synchronization.strategy", "sync" )
+				.tenants( TENANT_ID_1, TENANT_ID_2 )
+				.setup( IndexedEntity.class );
+
+		checkMultitenancy();
+	}
+
+	@Test
+	public void multiTenancyStrategy_enabledByMapping() {
+		sessionFactory = setupHelper.start()
+				.withProperty( "hibernate.search.automatic_indexing.synchronization.strategy", "sync" )
+				.tenants( TENANT_ID_1, TENANT_ID_2 )
+				.setup( IndexedEntity.class );
+
+		checkMultitenancy();
+	}
+
+	@Test
+	public void multiTenancyStrategy_none() {
+		try {
+			setupHelper.start()
+					.withProperty( "hibernate.search.backend.multi_tenancy.strategy", "none" )
+					.tenants( TENANT_ID_1, TENANT_ID_2 )
+					.setup( IndexedEntity.class );
+		}
+		catch (AssumptionViolatedException e) {
+			// do not catch this kind of exception
+			throw e;
+		}
+		catch (Exception e) {
+			assertThat( e ).isInstanceOf( SearchException.class )
+					.hasMessageMatching( FailureReportUtils.buildFailureReportPattern()
+							.typeContext( IndexedEntity.class.getName() )
+							.defaultBackendContext()
+							.failure( "Invalid backend configuration: " +
+									"index 'indexed' requires multi-tenancy but no multi-tenancy strategy is set" )
+							.build()
+					);
+		}
```

Comment:
You really should use `Assertions.assertThatThrownBy()`.

Reply:
Using `Assertions.assertThatThrownBy()` I got some errors if the dialect is not H2.
For instance [see this test execution](https://ci.hibernate.org/job/hibernate-search-personal-fabio/job/HSEARCH-4163/3/testReport/junit/org.hibernate.search.integrationtest.mapper.orm.realbackend.multitenant/RealBackendDatabaseMultitenancyIT(database-postgresql-default-elasticsearch)/Non_default_environments___database_postgresql___multiTenancyStrategy_none/).
I don't know if there exists a way to do that.

Reply:
I see.

You could try something like this:

```
assertThatThrownBy( ... )
        .asInstanceOf( rethrowAssertionFailures() )
        .isInstanceOf( ... )
        ....
```

And put this in `org/hibernate/search/util/impl/test/rule/AssertFactories.java`:

```
public final class AssertFactories {

    private AssertFactories() {
    }

    public static InstanceOfAssertFactory<Throwable, AbstractThrowableAssert<?, ? extends Throwable>> rethrowAssertionFailures() {
        return new InstanceOfAssertFactory( Throwable.class, throwable -> {
            if ( throwable instanceof org.junit.internal.AssumptionViolatedException
                        || throwable instanceof AssertionError ) {
                throw Throwables.toRuntimeException( throwable );
            }
            return Assertions.assertThat( throwable );
        } );
    }

}

```

If you have to stick with this `try`/`catch` for some reason, know that the problem is this test will pass if the startup doesn't throw an exception. So please at least add an Assertions.fail() at the end of the `try` block.

Reply:
If I try to implement the InstanceOfAssertFactory as you write above, I got this error:
> java.lang.ClassCastException: class org.assertj.core.api.ObjectAssert cannot be cast to class org.assertj.core.api.AbstractThrowableAssert (org.assertj.core.api.ObjectAssert and org.assertj.core.api.AbstractThrowableAssert are in unnamed module of loader 'app')

I'm going to add the `Assertions.fail()` at the end of the try block.

Reply:
That's just because of a missing diamond operator, I think:

```java
public final class AssertFactories {

    private AssertFactories() {
    }

    public static InstanceOfAssertFactory<Throwable, AbstractThrowableAssert<?, ? extends Throwable>> rethrowAssertionFailures() {
        return new InstanceOfAssertFactory<>( Throwable.class, throwable -> {
            if ( throwable instanceof org.junit.internal.AssumptionViolatedException
                        || throwable instanceof AssertionError ) {
                throw Throwables.toRuntimeException( throwable );
            }
            return Assertions.assertThat( throwable );
        } );
    }

}
```